Src/AutoNSubstitute/NSubstituteVirtualMethodsCommand.cs

```diff
@@ -172,40 +174,50 @@ public ISpecimenContext Context
                 get { return context; }
             }
 
-            public static void ReturnsForAnyArgs<T>(T value, Func<CallInfo, T> returnThis)
-            {
-                ReturnsForAnyArgsMethodInfo.Invoke(new object[] { value, returnThis });
-            }
-
             public static void Returns<T>(T value, Func<CallInfo, T> returnThis)
             {
                 ReturnsMethodInfo.Invoke(new object[] { value, returnThis });
             }
 
-            public void ReturnsUsingContext<T>(MethodInfo methodInfo)
+            public void ReturnsUsingContext(MethodInfo methodInfo)
             {
-                InvokeMethod(methodInfo);
-                ReturnsForAnyArgs<T>(default(T), callInfo =>
-                {
-                    var value = new Lazy<T>(() => (T)Context.Resolve(typeof(T)));
-                    object[] arguments = callInfo.Args();
-                    ReturnsFixedValue(methodInfo, callInfo, value);
-                    var returnValue = value.Value;
-                    InvokeMethod(methodInfo, arguments);
+                Substitute
+                    .WhenForAnyArgs(_ => InvokeMethod(methodInfo))
+                    .Do(callInfo =>
+                    {
+                        var arguments = callInfo.Args();
+                        var call = Tuple.Create(methodInfo, arguments.ToArray());
+                        if (methodCalls.Any(
+                            x =>
+                                call.Item1 == x.Item1 &&
+                                call.Item2.SequenceEqual(x.Item2)))
+                            return;
+                        methodCalls.Add(call);
+
+                        var value = Resolve(methodInfo.ReturnType);
+                        if (value is OmitSpecimen)
+                            return;
+
+                        ReturnsFixedValue(methodInfo, value);
+                        InvokeMethod(methodInfo, arguments);
+                    });
+            }
 
-                    return returnValue;
-                });
+            private object Resolve(Type type)
+            {
+                return Task.Factory
+                    .StartNew(() => Context.Resolve(type))
+                    .Result;
```

Comment:
@dcastro Not really.

When the mock is created, I set up a callback to be called when the method is called. This callback does not return the value. Instead, it sets up the method to return the value, or does not set up at all if it sees a ``OmitSpecimen``. 

To make a long story short, it will resolve the return value when the method is first called, but it may or may not set up the method to return the value.